### DISCUSSION TITLE
Lock statement patterns

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#94] [General]
# Lock statement patterns

A few weeks back, I wrote an article "A new lock type in .NET 9" where I showcased the new Lock type. Nothing fancy - well, at least it was more expressive. But now the dotnet team went a step further with this!

https://steven-giesel.com/blogPost/4cf6d48a-ec9d-4c68-961c-31fd8d8c1340

**Comment by Patrick-3000** [comment]
I have one question regarding the usage of `Monitor`: your write that

> The Lock class utilizes a Monitor under the hood.

As far as I can see from your lowered code, the `object`-Implementation also uses the `Monitor` class. So it seems to me, that in both ways the `Monitor` class is used. Is that true?

**Reply by linkdotnet** [reply]
Hupps that is on me - as seen in the code the new Lock class doesn’t utilise a Monitor.
I will fix that, thanks.

**Comment by manuel-ornato** [comment]
Thanks for this tip. Is there a tradeoff? Some context in which one can't use the Lock class?

**Reply by linkdotnet** [reply]
The biggest thing right now is that it is experimental for a reason. The implementation is quite complex.
For starters partially the new version uses spinning for waiters which may or may not improve performance (or make it worse).

Checking the source code, you can find also bits like this:
```csharp
// The lock is mostly fair to release waiters in a typically FIFO order (though the order is not guaranteed).
// However, it allows non-waiters to acquire the lock if it's available to avoid lock convoys.
//
// Lock convoys can be detrimental to performance in scenarios where work is being done on multiple threads and
// the work involves periodically taking a particular lock for a short time to access shared resources. With a
// lock convoy, once there is a waiter for the lock (which is not uncommon in such scenarios), a worker thread
// would be forced to context-switch on the subsequent attempt to acquire the lock, often long before the worker
// thread exhausts its time slice. This process repeats as long as the lock has a waiter, forcing every worker
// to context-switch on each attempt to acquire the lock, killing performance and creating a positive feedback
// loop that makes it more likely for the lock to have waiters. To avoid the lock convoy, each worker needs to
// be allowed to acquire the lock multiple times in sequence despite there being a waiter for the lock in order
// to have the worker continue working efficiently during its time slice as long as the lock is not contended.
//
// This scheme has the possibility to starve waiters. Waiter starvation is mitigated by other means, see
// TryLockBeforeSpinLoop() and references to ShouldNotPreemptWaiters.
```

That indicates that Microsoft is well aware of the changes they made and potentially can lead to worse performance than the `Monitor`. So for high-thorughput scenario you really need a testbed to measure the difference between `lock` with object or `lock` with the `System.Therading.Lock` type.

Also imagine scenarios where you want to inform the waiting queue via `Monitor.Pulse(_lock)` - this code relies on the `Monitor` itself.

For non-critical performance parts of the application my gut tells me it is fine to use the new type once its polished (and not behind a feature flag) - but time will tell.

Anyway - I will add this to the blog post itself! Thanks @manuel-ornato

**Comment by MarkCiliaVincenti** [comment]
Thanks for this!

I've created [a library](https://github.com/MarkCiliaVincenti/Backport.System.Threading.Lock) that helps backport it to pre .NET 9.0. It doesn't make it faster than locking on an object, but if you multi-target you can use the .NET 9.0 syntax and it will work just fine.

**Reply by linkdotnet** [reply]
Interesting! I like - .NET does this from time to time as well to make transition easier!
